### DISCUSSION TITLE
Streams replica can fail to recover and fail to accept consumers in case of an abrupt (`kill -9`-like) node process termination on Windows

### DISCUSSION THREAD

**Discussion by sysupbda** [rabbitmq/rabbitmq-server#12054] [Other] [answered]
### Describe the bug

When a single node RabbitMQ on Windows x64,  using OTP 26.2.5.2 and RabbitMQ v3.13.6, with a stream that is being published to and being consumed from, fairly often the recovery from crashes will result in an unavailable stream.

1- A node with just 50-100 clients, with just a couple of messages of a couple of kbs published to exchanges and streams per hour, can try to allocate 450Mb and fail (with erl crash dump file) despite 2Gb being available. I assume that there is no contiguous 450Mb available at that point. I have not nailed down what can cause a very little used RabbitMQ with few clients to allocate that much RAM.
(I don't know how to investigate or reproduce this one, but I am coachable)

2- Once RabbitMQ crashes because it could not allocate memory (because of (1)), or because it was killed at a particular moment, it sometimes missing an index file for which it has a segment file. These left over segments do not cause issues but become unmanaged resources and are never removed. In simple words, 00000000000000725366.segment is left there for ever.

![image](https://github.com/user-attachments/assets/f53c3177-e7ad-4003-925f-8c95914eb0e4)

3- The directory under %APPDATA%\db\streams\...\ also has an index file with zero bytes. This causes an **infinite loop** within the stream coordinator managing replicas (despite it being a single node cluster) which **keeps the stream unavailable** instead of recovering by skipping that ghost index file. 

![image](https://github.com/user-attachments/assets/f53c3177-e7ad-4003-925f-8c95914eb0e4)

4- The `missing_file` warning from the stream coordinator is warranted although it might be best to explicitly say that the stream is not available. It also repeatedly logs warnings with no retry delays which can cause self-inflicted Denial-Of-Services or poor log strategy tradeoffs. I am not sure if it is possible, but it would be better to only warn periodically instead of continuously.
```
...
2024-08-15 20:20:59.254000+08:00 [warning] <0.3811.0> rabbit_stream_coordinator: failed to get tail of member __static_data_17237229779233216 on rabbit@DESKTOP-R73UFKT in 2 Error: {case_clause,missing_file}
2024-08-15 20:20:59.258000+08:00 [warning] <0.3812.0> rabbit_stream_coordinator: failed to get tail of member __static_data_17237229779233216 on rabbit@DESKTOP-R73UFKT in 2 Error: {case_clause,missing_file}
2024-08-15 20:20:59.262000+08:00 [warning] <0.3813.0> rabbit_stream_coordinator: failed to get tail of member __static_data_17237229779233216 on rabbit@DESKTOP-R73UFKT in 2 Error: {case_clause,missing_file}
...
```
I created my own version of osiris_log.erl where I added the stacktrace to the file_size() call that fails. I hope it is helpful. This is how I know it is accessing the empty index file and failing on the non-existent corresponding stream file instead of recovering.
See attachment:

[log_with_modified_osiris_log_to_add_stack_trace_to_file_size_that_tries_to_open_stream_file.log](https://github.com/user-attachments/files/16639274/log_with_modified_osiris_log_to_add_stack_trace_to_file_size_that_tries_to_open_stream_file.log)

5- Sometimes when RabbitMQ is violently shutdown with `taskkill /f /im erl.exe` and a connected erl -node ... that uses `erlang:halt(2).`, instead of trying to recover from the ghost index file, it just rejects clients consumption of the stream (see `crash_instead_of_consume.txt`)

[crash_instead_of_consume.txt](https://github.com/user-attachments/files/16639277/crash_instead_of_consume.txt)

### Reproduction steps

0. Setup a Windows Server 2022, or use Windows 10.0.19045 Build 19045 or an AWS Windows_Server-2022-English-Full-Base-2024.07.10 - Microsoft Windows 2022 Datacenter edition t2.large (8 Gb 2 vcpu) hvm 64bit ENA enabled = true, root device: ebs, 40Gb gp2

1.  Install https://github.com/erlang/otp/releases/download/OTP-26.2.5.2/otp_win64_26.2.5.2.exe and https://github.com/rabbitmq/rabbitmq-server/releases/download/v3.13.6/rabbitmq-server-3.13.6.exe

2. Enable streams and friends with `rabbitmq-plugins.bat enable rabbitmq_stream_management`

3. Create a stream that will have frequent eval_retention calls. I attached a config that can be imported with `rabbitmqctl import_definitions definitions.json` and sets up administrator/administrator with [definitions.json](https://github.com/user-attachments/files/16639203/definitions.json)

4. publish and crash at the same time:

- Publish data to the stream that will result it retention evaluations 
(I am attaching this silly example to help publish and consume)
 [break_rabbit.zip](https://github.com/user-attachments/files/16639324/break_rabbit.zip)

- Run `taskkill /f /im erl.exe` while publishing

- Run `erlang:halt(2).` while connected to the cluster of one node

5. Repeat 4 until you find some of the described failure cases

### Expected behavior

Automated recovery from any failure scenario for streams.

Data in the mnesia and db subdirectories should be fully managed by RabbitMQ and no orphaned files should be left around, even when there was a failure case.

Logs need to remain relevant, even in failure cases.

### Additional context

I tried to nail down the root cause, and tried to make very reliable reproduction steps, but sadly I run into new issues every time I try to replicate. So I presented what I have observed, and the attempts to reproduce. Which of these issues you will see depends on luck apparently, but I think all these issues are unacceptable.

While I do regularly use Elixir, I have a limited experience of Erlang. So I can be coached to get additional information or modify code if I can get some support in doing so.

I have been using RabbitMQ for 10 years but am new to streams, and was more on the developer side than managing my own instances as an administrator. But I am fast learner. Feel free to point me to documentation and ask me to digest it.

If I can contribute in some way as a junior contributor to RabbitMQ, I am happy to support the team in the way that supports the team the best for these issues.

**Comment by michaelklishin** [comment]
Thank you for providing the details.

We will never promise "recovery from any failure scenario" because there scenarios where there is no path to recovery, or rather, in order to recover you must throw away some data.

Modern quorum queues already recover from many similar scenarios (in at least some cases, by deleting corrupt segment files 🤷), so streams can be more defensive.

**Comment by michaelklishin** [comment]
@sysupbda can you share a relevant directories from the node that runs into this "error logging loop"? Since you have a way to reproduce with test data.

**Comment by sysupbda** [comment]
Thank you @michaelklishin for your prompt answer.

I have no node where I can reproduce all issues. But I can share a node where I have the empty index and a missing segment issue (stream unavailable with infinite retry loop and log spamming), and the node where I get a consumption error but no other information.

Are these the directories that are most relevant? The %APPDATA%\RabbitMQ directory?

[RabbitMQ_stream unavailable with infinite retry loop empty index and missing segment.zip](https://github.com/user-attachments/files/16639623/RabbitMQ_stream.unavailable.with.infinite.retry.loop.empty.index.and.missing.segment.zip)

**Comment by sysupbda** [comment]
Here is the other one:

[RabbitMQ_with_segment_file_laying_around_and_rejecting_consumption_from_the_stream.zip](https://github.com/user-attachments/files/16639630/RabbitMQ_with_segment_file_laying_around_and_rejecting_consumption_from_the_stream.zip)

**Comment by kjnilsson** [comment]
@sysupbda thanks for providing the data directories and I have confirmed there are a couple of issues in the code that I am working to fix.

However my main concern is the behaviour of windows in this case. From your description it appears that a simple process crash/forced exit is enough to lose page cache data (or file cache, cache manager cache, whatever it is called on windows).

We tested similar cases heavily on linux and at no point did we lose page cache data from a process crash (as that isn't how the page cache works). To lose page cache data we had to force terminate the machine or vm.

I just can't believe that is how windows works so there must be something else going on this case but we test very little on windows and don't have as much familiarity with it. _If_ that is how windows is designed to work w.r.t to unflushed disk data then I don't think windows should be used for production systems that use RabbitMQ streams. At least not without further OS specific changes.

**Comment by sysupbda** [comment]
Hi Karl !

Thank you for your fast and detailed reply!

I understand and agree with your concerns.

I had in mind to support the project with two fixes that I believe would
take us a very long way:

1- a stage before streams are loaded where the environment is "recovered".

If we did this, we could eventually maybe share recovery best practices
between quorums and streams.

I would categorize them in to three groups:
A) can self-heal, and I think nearly all the issues I listed can be solved
by this. Essentially we start by validating our storage and if recovery is
possible without any intervention, we recover.
=> I think we already have a process to copy from other replicas. Do we
maybe just need to add the check whether any 0 byte index files exist
without segment file?

B) can be healed but needs the owner of the system to make some decisions.
One such decision might be to abandon a subset of data that was not
recoverable.
=> I dont think that any of the issues I mentioned require this. Maybe it
could be done at a later stage?


2- a process that runs as a service and cleans up orphaned, useless segment
files that have no corresponding index file. This can happen with latency,
but probably should happen.


I had no time to dive into it, but I wanted to discuss it on the mailing
list. I wasn't sure whether this issue or the mailing list was the best
place.

What do you think?

Is there any way I can support the process?

Thanks!

**Reply by sysupbda** [reply]
@kjnilsson , concerning your main concern about caching and lost page cache data, is this something we still want to investigate? Is that something I could contribute to?

**Comment by kjnilsson** [comment]
> Is there any way I can support the process?

The most helpful thing now would be to share some insight into how windows works w.r.t to page cache. Are these services running in some kind of sandboxing/isolation mode that could cause a simple process crash to lose unsynced file caches?

**Reply by sysupbda** [reply]
I see. I would have to learn about how Windows handles page caches myself too. This might take me some time. Can I come back to you on that? 


The data corruption I experienced was reproduced on three different environments:
My home PC: Windows 10 pro, typical desktop setup, 

An AWS windows server t2.large with hvm on a ebs disk 

A windows server VM that my client, a bank, runs and to which I only have very limited access. It might be running VMware but honestly I don't know and they are not sharing too much information. 

As a side note :
I chose RabbitMQ because it of its features. However a big reason was that it advertises that it runs on Windows. The choice of systems that officially support Windows (and can be run in-premises) is thin. I thought erlang had a special trick up it's sleaves but I guess Erlang is not magical. It might make sense to officially not support Windows until a battery of tests are created for it?

**Comment by kjnilsson** [comment, marked as answer]
@sysupbda I have made some changes that should at least handle the two cases you reported. Please can you test this?

https://github.com/rabbitmq/rabbitmq-server/pull/12073

**Comment by sysupbda** [comment]
Hi Karl,
Thank you for the prompt update! I am setting up a development environment on the AWS EC2 instance I used to reproduce the issue. I will let you know if I can still reproduce when the new version of osiris is used.

PS: I am going to naively do a `git checkout osiris-1.8.3` on the `rabbitmq-server` repo. I hope that is what you are also expecting from me.

Edit: I see it is a branch, not a tag. Will try to run it that way. I only used installers until today.

**Reply by sysupbda** [reply]
Sorry @kjnilsson for not figuring it out on my own, but I am having trouble setting up with your branch. Any chance you have an easy answer for this? I am attempting this on my AWS EC2 Windows Server machine I used to replicate the original issues. I am happy to learn, so you can also point me in the right direction and I will head there.

Once I setup git, checked out the rabbitmqserver repo, switched to the osiris-1.8.3 branch, I attempted:
1) to follow the instructions of PKG_WINDOWS.md. Everything worked fine until it asked me to checkout an archived repo. At that point, I assumed it was maybe outdated? Should I file a bug for this or even contribute a pull request to have that file removed?

2) to follow the instructions of BAZEL.md for Windows, but sadly it fails to patch something?

```
PS C:\Users\Administrator\rabbitmq-server> bazel clean
INFO: Starting clean.
PS C:\Users\Administrator\rabbitmq-server> bazel run broker
WARNING: For repository 'platforms', the root module requires module version platforms@0.0.8, but got platforms@0.0.9 in the resolved dependency graph.
INFO: Repository rules_erlang~~erlang_package~cowboy instantiated at:
  <builtin>: in <toplevel>
Repository rule hex_archive defined at:
  C:/users/administrator/_bazel_administrator/t3e7dvyl/external/rules_erlang~/hex_archive.bzl:177:30: in <toplevel>
INFO: repository @@rules_erlang~~erlang_package~cowboy' used the following cache hits instead of downloading the corresponding file.
 * Hash '8a7abe6d183372ceb21caa2709bec928ab2b72e18a3911aa1771639bef82651e' for https://repo.hex.pm/tarballs/cowboy-2.12.0.tar
If the definition of 'repository @@rules_erlang~~erlang_package~cowboy' was updated, verify that the hashes were also updated.
ERROR: An error occurred during the fetch of repository 'rules_erlang~~erlang_package~cowboy':
   Traceback (most recent call last):
        File "C:/users/administrator/_bazel_administrator/t3e7dvyl/external/rules_erlang~/hex_archive.bzl", line 85, column 10, in _impl
                patch(ctx)
        File "C:/users/administrator/_bazel_administrator/t3e7dvyl/external/bazel_tools/tools/build_defs/repo/utils.bzl", line 228, column 21, in patch
                fail("Error applying patch command %s:\n%s%s" %
Error in fail: Error applying patch command rm ebin/cowboy.app:
java.io.IOException: ERROR: src/main/native/windows/process.cc(202): CreateProcessW("bash" -c "rm ebin/cowboy.app"): The system cannot find the file specified.
 (error: 2)
ERROR: no such package '@@rules_erlang~~erlang_package~cowboy//': Error applying patch command rm ebin/cowboy.app:
java.io.IOException: ERROR: src/main/native/windows/process.cc(202): CreateProcessW("bash" -c "rm ebin/cowboy.app"): The system cannot find the file specified.
 (error: 2)
ERROR: C:/users/administrator/rabbitmq-server/deps/rabbitmq_web_stomp/BUILD.bazel:53:13: //deps/rabbitmq_web_stomp:erlang_app depends on @@rules_erlang~~erlang_package~cowboy//:erlang_app in repository @@rules_erlang~~erlang_package~cowboy which failed to fetch. no such package '@@rules_erlang~~erlang_package~cowboy//': Error applying patch command rm ebin/cowboy.app:
java.io.IOException: ERROR: src/main/native/windows/process.cc(202): CreateProcessW("bash" -c "rm ebin/cowboy.app"): The system cannot find the file specified.
 (error: 2)
ERROR: Analysis of target '//:broker' failed; build aborted: Analysis failed
INFO: Elapsed time: 6.761s, Critical Path: 0.06s
INFO: 1 process: 1 internal.
ERROR: Build did NOT complete successfully
ERROR: Build failed. Not running target
FAILED:
    Fetching repository @@rules_erlang~~erlang_package~ranch; starting
    Fetching C:/users/administrator/_bazel_administrator/t3e7dvyl/external/rules_erlang~~erlang_package~ranch; Extracting contents.tar.gz
```
Edit: I originally made no changes. And after reviewing the `user.bazelrc` I corrected the logs value to `build --@rules_erlang//:ct_logdir=/C:/Users/Administrator/rabbitmq-server/logs`

Not sure if it helps, but the context is that I copied the user-template.bazelrc to user.bazelrc and only changed the last line. All the other files are just as they were in the repo for the osiris-1.8.3 branch. For example, in MODULE.bazel, I have:
```
...
bazel_dep(
    name = "rules_erlang",
    version = "3.16.0",
)
...
erlang_config = use_extension(
    "@rules_erlang//bzlmod:extensions.bzl",
    "erlang_config",
)

use_repo(
    erlang_config,
    "erlang_config",
)
...
erlang_package = use_extension(
    "@rules_erlang//bzlmod:extensions.bzl",
    "erlang_package",
)
...
erlang_dev_package = use_extension(
    "@rules_erlang//bzlmod:extensions.bzl",
    "erlang_package",
)
```

Sadly I am unfamiliar with bazel and the RabbitMQ build process, so it would be a struggle to go further without help.

3) I tried to use `make` instead of bazel but I dont think our makefiles in rabbitmq's repo work for windows. I get this:
```
PS C:\Users\Administrator\rabbitmq-server> make
process_begin: CreateProcess(NULL, elixir --eval ":io.format \"~s~n\", [:code.lib_dir :elixir ]", ...) failed.
Makefile:50: pipe: No error
sub(/\)$/ was unexpected at this time.
main was unexpected at this time.
-f was unexpected at this time.
process_begin: CreateProcess(NULL, uname -s, ...) failed.
erlang.mk:74: pipe: No error
process_begin: CreateProcess(NULL, uname -o, ...) failed.
erlang.mk:92: pipe: No error
erlang.mk:95: *** Unable to detect platform. Please open a ticket with the output of uname -a..  Stop.
```

I recognize `uname` as a linuxy/macosy kind of command.

**Reply by sysupbda** [reply]
Some extra info, I also tried to run `make` within `msys` but it can't find elixir despite it being installed on the machine with choco and not being available to be installed with `pacman`.

**Reply by kjnilsson** [reply]
ok thanks, let me see if I can produce a windows artefact for you

**Reply by kjnilsson** [reply]
[rabbitmq-server-4.1.0+alpha.140.gbaa6410.zip](https://github.com/user-attachments/files/16709287/rabbitmq-server-4.1.0%2Balpha.140.gbaa6410.zip)

Can you try this?

**Reply by sysupbda** [reply]
Thanks Karl. I just tried on that EC2 instance. Sadly it is missing the escript subdirectory and I only slept 3 hours last night.

```
C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410\sbin>set RABBIT
RABBITMQ_HOME=C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410

C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410\sbin>set PATH
Path=C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Amazon\cfn-bootstrap\;C:\Program Files\dotnet\;C:\Program Files\Microsoft VS Code\bin;C:\Program Files\Git\cmd;C:\ProgramData\chocolatey\bin;C:\Users\Administrator\AppData\Local\Microsoft\WindowsApps;C:\Users\Administrator\.dotnet\tools;C:\tools\msys64
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC

C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410\sbin>dir C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410\escript
 Volume in drive C has no label.
 Volume Serial Number is 987C-EBFE

 Directory of C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410

File Not Found

C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410\sbin>rabbitmqctl.bat
escript: no such file or directory: 'C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410\escript\rabbitmqctl'
```

Instead of calling the sbin commands to run rabbitmq, I thought you might be expecting me to run bazel, so I attempted:
1) Running `bazel run` or `bazel sync`
2) Adding a `user.bazelrc` and then running `bazel run` or `bazel sync`

In both cases, I get:
```
C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410>bazel run
ERROR: Error computing the main repository mapping: Every .bzl file must have a corresponding package, but '//deps/amqp10_client:activemq.bzl' does not have one. Please create a BUILD file in the same or any parent directory. Note that this BUILD file does not need to do anything except exist.
Computing main repo mapping:

C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410>bazel sync
DEBUG: Rule 'rules_erlang' indicated that a canonical reproducible form can be obtained by modifying arguments commit = "92c6f39bde010a0651935774eb241533aa16e49c" and dropping ["tag"]
DEBUG: Repository rules_erlang instantiated at:
  C:/users/administrator/downloads/rabbitmq-server-4.1.0+alpha.140.gbaa6410/rabbitmq-server-4.1.0+alpha.140.gbaa6410/WORKSPACE:19:15: in <toplevel>
Repository rule git_repository defined at:
  C:/users/administrator/_bazel_administrator/dnviwdod/external/bazel_tools/tools/build_defs/repo/git.bzl:189:33: in <toplevel>
ERROR: error loading package '': Every .bzl file must have a corresponding package, but '//deps/amqp10_client:activemq.bzl' does not have one. Please create a BUILD file in the same or any parent directory. Note that this BUILD file does not need to do anything except exist.

C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410>dir deps
 Volume in drive C has no label.
 Volume Serial Number is 987C-EBFE

 Directory of C:\Users\Administrator\Downloads\rabbitmq-server-4.1.0+alpha.140.gbaa6410\rabbitmq-server-4.1.0+alpha.140.gbaa6410\deps

08/22/2024  02:11 PM    <DIR>          .
08/22/2024  02:20 PM    <DIR>          ..
08/22/2024  02:11 PM    <DIR>          .hex
08/22/2024  02:11 PM    <DIR>          accept
08/22/2024  02:11 PM    <DIR>          amqp10_common
08/22/2024  02:11 PM    <DIR>          amqp_client
08/22/2024  02:11 PM    <DIR>          cowboy
08/22/2024  02:11 PM    <DIR>          eetcd
08/22/2024  02:11 PM    <DIR>          rabbitmq_amqp1_0
08/22/2024  02:11 PM    <DIR>          rabbitmq_auth_backend_http
08/22/2024  02:11 PM    <DIR>          rabbitmq_auth_backend_oauth2
08/22/2024  02:11 PM    <DIR>          rabbitmq_cli
08/22/2024  02:11 PM    <DIR>          rabbitmq_codegen
08/22/2024  02:11 PM    <DIR>          rabbitmq_federation
08/22/2024  02:11 PM    <DIR>          rabbitmq_recent_history_exchange
08/22/2024  02:11 PM    <DIR>          rabbitmq_shovel
08/22/2024  02:11 PM    <DIR>          rabbitmq_shovel_prometheus
08/22/2024  02:11 PM    <DIR>          rabbitmq_web_dispatch
08/22/2024  02:11 PM    <DIR>          rabbitmq_web_mqtt
08/22/2024  02:11 PM    <DIR>          rabbitmq_web_stomp_examples
08/22/2024  02:11 PM    <DIR>          ranch
08/22/2024  02:11 PM    <DIR>          syslog
```

I could get some rest and try tomorrow to cross-compile a Windows installer from Linux or Mac and get this working that way. Alternatively, I was thinking that I could just recompile all `.erl` files into `.beam` and copy them into the main installation subdirectory, just like I did to help report the bug. It's kind of a workaround, though. If you want to point me to the correct documentation to build rabbitmq for Windows, I would try to follow those steps.

Either way, I will try it with the `.beam` files overwriting tomorrow with your new version of Osiris. But I really should get some rest right now.

**Reply by michaelklishin** [reply]
Now that https://github.com/rabbitmq/rabbitmq-server/pull/12073 is merged and soon enough will be its backport to `v4.0.x`, there will be a new alpha build on https://github.com/rabbitmq/rabbitmq-server-binaries-dev/releases available in some 2-3 hours.

**Reply by michaelklishin** [reply]
A few 3.13.7-alpha builds were produced earlier today, with at least one more coming in a couple of hours to https://github.com/rabbitmq/rabbitmq-server-binaries-dev/releases.

**Reply by sysupbda** [reply]
Thank you Michael, I ended up doing the tests with the 4.0 alpha release binary and the 3.13.6 + .beam files patches for the osiris plugin.

@kjnilsson , I can only reproduce one of the issues I had reported. As a recap:

1) 450Mb of contiguous memory being allocated at once is not great, but it turns out a junior dev had a memory leak in their code that created a connection, channel and consumer every 10 seconds I guess. So having to allocate contiguous 450 Mb after a week of such abuse is actually impressively low. We can remove that from our list of interesting points.

2) After applying the patch of `.beam` files on 3.6, I could see in the logs that the orphaned segment was removed and it was noted in the logs.

3) I could not reproduce the empty index file without matching segment but also did not find anything in the logs about it. It was quite easy to reproduce before the patch so my guess is that this was fixed. I could not reproduce with the patched 3.6 nor with the 4.0 alpha.

4) Because we no longer have the problematic index file, we also can't reproduce the log abuse.

5) The consumption being broken is a complicated one. I could reproduce it using 3.6 patched and 4.0 alpha (see attachments). That being said, it only gets in this state in the first place if you can reproduce point (3). I still think it should be fixed, but it no longer would affect me and in the short term it would not affect anyone as far as I can tell.

![wont_allow_consumption](https://github.com/user-attachments/assets/f2253b44-ce9d-4a29-b84c-8b6ce83540ac)


[RabbitMQ_still_rejecting_consumption_from_the_stream.zip](https://github.com/user-attachments/files/16721518/RabbitMQ_still_rejecting_consumption_from_the_stream.zip)
[break_rabbit.zip](https://github.com/user-attachments/files/16721519/break_rabbit.zip)

**Reply by sysupbda** [reply]
Can you help me understand what criteria will be applied to release the official bug fix release for this as a v3.13.7 (no-alpha)?

**Reply by michaelklishin** [reply]
@sysupbda if your question is when can `3.13.7` ship, the answer is it can ship with these improvements basically any day. The goal of alpha releases is to let contributors (even if not with code) like you to test out new changes.

**Reply by sysupbda** [reply]
You answered my question. I will be keeping an eye on the release of 3.13.7 and ask my client to upgrade to it when it is shipped.

I will also try to find how to contribute a coffee/beer to Karl and yourself. Not only because of the code contributions. But also because you called me a contributor. ;-)

**Reply by kjnilsson** [reply]
@sysupbda thanks for testing this. For point 5 yes that still remains and it is a curious one I have to say, I suspect that in a clustered environment this would self heal but again I'm confounded by the windows behaviour here.

What you are seeing here is an empty stream but with a non zero starting index. This is not a valid state and could only be caused by very low retention settings and loss of page cache data. What retention settings are you using for your test?

**Reply by sysupbda** [reply]
Hi Karl! I had originally used the following to create the situation. I only had replicated that case on my AWS EC2 instance I had purposefully created to replicate reliably:

```
{"arguments":{"x-max-age":"1m","x-max-length-bytes":40000,"x-queue-type":"stream","x-stream-max-segment-size-bytes":4000},"auto_delete":false,"durable":true,"name":"static_data","type":"stream","vhost":"/"}
```

I purposefully set the values quite low to force rabbit to create more segments and have more segments to delete.

The only remaining concern on that one is that it does not self-heal in that scenario. At least not in a single-node setup (when the stream is not replicated).

As I learn more about Erlang and testing such complicated distributed, multi-process scenarios, I might be able to create test cases to help replicate it. I am sorry I am not able to do that at this time, if it is possible to do at all.

**Reply by michaelklishin** [reply]
A stream that's limited to 40 kB? That's below one segment worth.

**Reply by sysupbda** [reply]
In this particular config, is the segment size not set to 4k and the stream to 40k? Did I misunderstand that lart? Or is that not supposed to work?

**Reply by michaelklishin** [reply]
It's not a configuration that makes much sense, even though it may be good for reproducing certain edge cases.

Streams are meant to be used for reasonably long storage of shared data sets that should be read repeatedly. A 40 kB limit is hardly very practical for a stream (or even a queue). RabbitMQ defaults assume multi-megabyte segment size.

Of course, if these were set specifically to reproduce what was reported in this discussion, then you can disregard my comment. Otherwise why not stick to stream defaults and a limit of some 128 MB, for example?

**Reply by sysupbda** [reply]
Thank you @michaelklishin for highlighting the concern. I strongly welcome input, particularly from the team that has a lot more experience with stream that I have. I did indeed choose a very small segment and overall stream size, in the hope that it should work. My architecture involves a variety of teams, and needed a strong decoupling of each component. Conceptually, this particular stream represents a kind of config that may be updated by the customer and nearly immediately reflected in all the workers (consumers of the queue).

If using a stream this way makes sense, even if it is not a common use case for it, I would argue that having the stream be small and the consumers be allowed to start at "offset 1" at startup. I felt it would add quite a lot of complexity for no benefit to start storing the offset to read from on startup. Most consumers are stateless. I might have missed an alternative solution here?

Is my use case a poor fit for streams? If it is an acceptable fit, would using larger segment/stream sizes not result in a longer startup time for every consumer rereading from offset 1? I was also worried it would create a sudden spike in networking that could ripple into other side effects that are hard to investigate for colocated systems.

**Reply by michaelklishin** [reply]
@sysupbda `3.13.7` has shipped https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.13.7. Thank you for reporting this recovery scenario and helping verify the fixes.

**Reply by sysupbda** [reply]
Thank you! I am extremely thankful!
